_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,228 @@
+
+The same is true for control-flow integrity (CFI).
+CFI defines a program as a directed graph of blocks of instructions with arcs in between them representing things like function calls, returns, and so on.
+Again, the existence of this graph is a property of the language.
+For example, in C++ if you call a virtual function then the object on which it's called must be a subclass of the class on which the function is defined.
+This set of properties can be quite rich and so most CFI schemes focus on preventing particularly dangerous invalid transitions, rather than preventing *all* invalid transitions.
+For example, a C CFI scheme may allow you to call `system` when you meant `puts`, because they take the same argument type, but prevent you from calling `fputs`.
+This is a dangerous conflation (it may allow an attacker to run a program rather than printing a string) but it doesn't corrupt the state of the running program.
+
+
+## You're in my language and you'll follow my language's abstract machine's rules!
+
+So what's the point of CHERI if an ISA can't give you these properties?
+CHERI is not designed to give you an object model that you must conform to, it's designed to *give language implementers tools to enforce these properties*.
+This is very different from a lot of earlier memory-safe system.
+Intel's [iAPχ 432](https://en.wikipedia.org/wiki/Intel_iAPX_432) was designed around the Algol model (though not very well), as was the [B5500](https://en.wikipedia.org/wiki/Burroughs_Large_Systems#B5500) (somewhat better).
+Various [Lisp machines](https://en.wikipedia.org/wiki/Lisp_machine) implemented the Lisp memory model.
+Such an approach is not feasible today, in a world where most programs use components written in a variety of languages.
+
+
+That's a very important distinction and it becomes even more important when you ask the follow-on question: Against whom, exactly, are these properties being enforced?
+
+We assume that a compiler has some notion of an object model.
+It may be able to enforce that memory model entirely because the source language has strong typing guarantees.
+For example, we have a JavaScript VM in the CHERIoT RTOS repository that builds its own garbage-collected type-safe object model using 16-bit integers to represent object pointers (more on that later).
+That compiler needs to generate binaries that interoperate with other code.
+The other code may be third-party code compiled with the same compiler but [rely on exploiting compiler bugs to bypass certain guarantees](https://github.com/Speykious/cve-rs).
+It may be written in a different language with different guarantees, or it may be assembly code with no enforced concept of types or an object model.
+Our goal is to allow the compiler for one language to enforce these properties against all of that other code, irrespective of the source language or compiler (if any) used to produce the other code.
```

Comment:
I like this characterization, but this is more general than just the compiler (it includes the ABI and the ISA). Abstractly, you want to allow developers to enforce these properties on their code against all of that other code irrespective of ... the other code.